docs/docsite/rst/dev_guide/testing_units_modules.rst

```diff
@@ -0,0 +1,534 @@
+These states are then used as returns from a mock object to ensure that the await function
+waits through all of the states that would mean the RDS instance has not yet completed
+configuration::
+
+   rds_i.await_resource(rds_client_double, "some-instance", "available", mod_mock,
+                        await_pending=1)
+   assert(len(sleeper_double.mock_calls) > 5), "await_pending didn't wait enough"
+
+By doing this we check that the await function will keep waiting through
+various strange states that it would be impossible to reliably trigger through the
+integration tests but which happen unpredictably in reality.
+
+Defining a module to work against multiple API versions
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+This case is especially important for modules interacting with many different versions of
+software; for example, package installation modules which might be expected to work over
+many different operating system versions.
+
+By using previously stored data from various versions of an API we can ensure that the
+code is tested against the actual data which will be sent from that version of the system
+even when the version is very obscure and unlikely to be available during testing.
+
+Ansible special cases for unit testing
+======================================
+
+There are a number of special cases for unit testing the environment of an Ansible module.
+The most common are documented below, and suggestions for others can be found by looking
+at the source code of the existing unit tests or asking on the Ansible IRC channel or mailing
+lists.
+
+Module argument processing 
+--------------------------
+
+There are two problems with running the main function of a module.  
+
+* It can be difficult to set up the arguments correctly so that the module will get them
+  as parameters.
+* All modules finish by calling either the ``module.fail_json`` or
+``module.exit_json``, but these won't work correctly in a testing environment.
+
+Passing Arguments
+-----------------
+
+To pass arguments to a module correctly, use a function that stores the
+parameters in a special string variable.  Module creation and argument processing is
+handled through the AnsibleModule object in the basic section of the utilities.  Normally
+this accepts input on ``STDIN``, which is not convenient for unit testing. When the special
+variable is set it will be treated as if the input came on ``STDIN`` to the module.::
+
+    import json
+    from ansible.module_utils._text import to_bytes
+
+    def set_module_args(args):
+        args = json.dumps({'ANSIBLE_MODULE_ARGS': args})
+        basic._ANSIBLE_ARGS = to_bytes(args)
+
+    simply call that function before setting up your module
+
+        def test_already_registered(self):
+            set_module_args({
+                'activationkey': 'key',
+                'username': 'user',
+                'password': 'pass',
+            })
+
+Handling exit correctly
+-----------------------
```

Comment:
Keep an eye on https://github.com/ansible/ansible/pull/31456 as it will affect the recommendation here.